src/TimeoutExtender.js

```diff
@@ -0,0 +1,163 @@
+
+'use strict'
+
+/**
+ * The number of seconds to place the API call to change the VisibilityTimeout in advance
+ * of its expiration time.
+ * @type {number}
+ */
+const API_CALL_LEAD_MS = 5000
+
+/**
+ * Option defaults.
+ * @type {Object}
+ */
+const optDefaults = {
+  visibilityTimeoutSecs: 30,
+  noExtensionsAfterSecs: 0
+}
+
+/**
+ * The TimeoutExtender is a module that attaches itself to a Squiss instance via event
+ * listeners, and uses the instance's public API to automatically extend the
+ * VisibilityTimeout of the message until it is either handled (which includes actions
+ * such as deleting or releasing the message) or it reaches an age that is greater than
+ * the `noExtensionsAfterSecs` option. Functionally, this class could be distributed as a
+ * third party module and attached to a Squiss instance by the user, however the common
+ * and convenient use case is for this class to be used internally by Squiss itself.
+ *
+ * For efficient operation, the TimeoutExtender combines a doubly-linked list with a
+ * hash map, keeping an index of every message while not incurring the RAM or CPU overhead
+ * of having to re-create an array every time a message is deleted from the middle of a
+ * sorted queue. Every operation in this class has a time complexity of O(1) and a space
+ * complexity of O(n).
+ */
+class TimeoutExtender {
+
+  /**
+   * Creates a new TimeoutExtender.
+   * @param {Squiss} squiss The Squiss instance on which this extender should operate
+   * @param {Object} opts
+   */
+  constructor(squiss, opts) {
+    this._opts = Object.assign({}, optDefaults, opts)
+    this._head = null
+    this._tail = null
+    this._index = {}
+    this._timer = null
+    this._squiss = squiss
+    this._squiss.on('handled', msg => this.deleteMessage(msg))
+    this._squiss.on('message', msg => this.addMessage(msg))
+    this._visTimeout = this._opts.visibilityTimeoutSecs * 1000
+    this._stopAfter = this._opts.noExtensionsAfterSecs * 1000
+  }
+
+  /**
+   * Adds a new message to the tracker.
+   * @param {Message} message A Squiss Message object
+   */
+  addMessage(message) {
+    const now = Date.now()
+    this._addNode({
+      message,
+      receivedOn: now,
+      timerOn: now + this._visTimeout - API_CALL_LEAD_MS
+    })
+  }
+
+  /**
+   * Deletes a message from the tracker, if the message is currently being tracked.
+   * @param {Message} message A Squiss Message object
+   */
+  deleteMessage(message) {
+    const node = this._index[message.raw.MessageId]
+    if (node) this._deleteNode(node)
+  }
+
+  /**
+   * Adds a message wrapper node to the linked list and hash map index.
+   * @param {{message: Message, receivedOn: number, timerOn: number}} node The node
+   * object to be added
+   * @private
+   */
+  _addNode(node) {
+    this._index[node.message.raw.MessageId] = node
+    if (!this._head) {
+      this._head = node
+      this._tail = node
+      this._headChanged()
+    } else {
+      this._tail.next = node
```

Comment:
ahahahahaha I was laughing to myself the whole time. If someone ever asks "Why would I need this in real life?" I'm linking them here.